extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/ArcProxyBeanMetaDataClassNormalizer.java

```diff
@@ -6,16 +6,21 @@
 
 /**
  * In the case of a proxy generated by Arc, return the parent class.
+ * <br />
+ * This can have more than one level of hierarchy, in example:
+ * <ul>
+ * <li>When using @{@link InjectMock} or @{@link InjectSpy}</li>
+ * </ul>
  */
 public class ArcProxyBeanMetaDataClassNormalizer implements BeanMetaDataClassNormalizer {
 
     @Override
     public <T> Class<? super T> normalize(Class<T> beanClass) {
-        if (Subclass.class.isAssignableFrom(beanClass)) {
-            return beanClass.getSuperclass();
+        if (!Subclass.class.isAssignableFrom(beanClass)) {
+            return beanClass;
         }
 
-        return beanClass;
+        return normalize(beanClass.getSuperclass());
```

Comment:
Any chance you could implement it with a loop rather than recursion? I agree that we should have things nested too much but it's good practice to avoid recursion when you can.

Reply:
I was avoiding loops because it looked uglier.

In this case, in don't see I scenario where the recursion could go bad, but sure, I'll rewrite it.

Reply:
Yeah, I agree it's uglier but that's a good habit to have.